### ISSUE TITLE
When rendering with option --save_section the rendered file names are appended with 0001, 0002, 0003 instead of the "name" attribute.

### ISSUE THREAD

**Issue by ubaldot** [ManimCommunity/manim#3471]
I am not sure if this shall be categorized as a bug, so let me know if I have to move it elsewhere.

As per title, when rendering with option `--save_section `the rendered file names are appended with 0001, 0002, 0003 instead of the "name" attribute.

for example, say that iI have something like the following:

```
class MyScene(Scene)
    def construct(self):
          self.next_section("MyFirstSection")
          # some code
          self.next_section("Section_Two")
          # some other code
```

and then I render with `--save-sections` option‚ then the rendered files have names:

```
MyScene_0001.mp4
MyScene_0002.mp4
```

whereas it would be nicer if they inherit the section name, e.g.

```
MyScene_MyFirstSection.mp4
MyScene_Section_Two.mp4
```

**Comment by behackl** [comment]
Not sure I agree. The name could perhaps appear somewhere in the filename, but replacing the numbering completely leads to unsorted files in the file explorer.

**Comment by ubaldot** [comment]
> The name could perhaps appear somewhere in the filename

That could be an idea! 
But I am wondering, why it would be a problem to have unsorted files? 
At the end, even when you have a file with multiple scenes you would have them unsorted. 
For example:

```
class Foo(Scene):
    def construct(self):

class Bar(Scene):
    def construct(self):   

class FooBar(Scene):
    def construct(self):  
```

when you render all of them you would have them unsorted, no? :)

**Comment by ubaldot** [comment]
> Because if you read them back from disk it becomes a problem of sorting the files by name

This sounds like a more a developer issue than a user issue.

**Comment by doaamuham** [comment]
I would like to take on this issue

**Comment by MrDiver** [comment]
> This sounds more like a developer issue than a user issue.

It is still a user issue because if you drag the section files (with no numbering) into a video editor then they will appear sorted by name, because lexically sorting files is a standard in many software projects.

So by removing the numbers, you force the user to add the files one by one and sort them manually in a video editor, which is way more work than it needs to be.

Summary: It is indeed a user problem, that's why we have the numbers!

**Comment by behackl** [comment]
We could just do both, `SceneName_num_SectionName`, as @chopan050 has already suggested on the corresponding PR linked above.

As section names do not need to be valid identifiers, users will have to avoid naming their sections in a way that results in illegal characters in file names.

**Comment by ubaldot** [comment]
…or do something like [this](https://stackoverflow.com/questions/1323364/in-python-how-to-check-if-a-string-only-contains-certain-characters) perhaps?

**Comment by chopan050** [comment]
There are things that need to be clarified, so that this discussion can advance. Please make sure, @ubaldot, that you fully understand them thoroughly.

There are misunderstandings in mainly 2 topics:
1. The differences between a scene and a section, and when to use each one (leading to a _false equivalence_).
2. Which ideas are really being argued against / in favor of, and which are not (leading to a _straw man fallacy_).

## 1. The differences between a scene and a section

> why it would be a problem to have unsorted files?
At the end, even when you have a file with multiple scenes you would have them unsorted.

Apples and oranges:
- Scenes don't have a logical/temporal order ingrained into them. Unless you define your own Scene subclasses or something similar...
  - ...your scenes aren't really connected to each other / don't really depend one on the other, in the "programming / data sequence" sense.
  - ...you can change the order of your scenes in your file, and nothing happens with the rendering. The resulting videos/images are the same.
  - Therefore, **they stand on their own**. That doesn't mean they must represent a whole project, like `ComplexNumbers`: something smaller, a subtopic like `DefinitionOfI`, where only the specific succession of events/animations/objects necessary to explain the number i, is needed. No other Mobjects, no other animations: `DefinitionOfI` can be self-contained.
- Sections, **by definition**, are chronologically ordered.
  - They consist of "snapshots" of scenes, they're based on saved states **at specific times** during your scene.
  - Usually, what happens in a section depends on what happened in the previous section(s). If my section N° 2 only contains `self.play(Rotate(square))`, that only really makes sense if I previously added that square in section N° 1, maybe via `self.play(FadeIn(square))`.
  - Mess with the order, and you'll likely mess up your scene.
  - If a section does not depend on any of the previous sections of your scene... that's a signal for you to consider creating a new scene for it, because maybe you're unnecessarily packing your whole complex project inside a same scene. Divide to simplify things.
  - 

TL;DR
- Sections inside a scene are meant to represent a sequence of interdependent events. Sections depend on each other. They aren't really able to represent a "topic" on their own, they're inherently numbered fragments which need to be assembled in a specific order into something coherent.
- Scenes are more meant to be standalone products which describe a certain topic or subtopic. They can be part of a bigger project, sure, but they don't need to.

Therefore, they can't be really compared in this sense (false equivalence). While it is OK to have unsorted scenes (because they can function as standalone projects), we cannot really say the same about sections in the same scene.

For the layout you described:
```
ComplexNumbers_cartesian_representation.mov
ComplexNumbers_polar_representation.mov
ComplexNumbers_conversions.mov
```
it seems that those sections could've perfectly been **their own, standalone scenes** instead. All these scenes could be defined in a single `complex_numbers.py` file:
```py
# complex_number.py
class CartesianRepresentation(Scene):
    # ...
class PolarRepresentation(Scene):
    # ...
```
leading to a layout like this instead:
```
CartesianRepresentation.mov
PolarRepresentation.mov
Conversions.mov
```
Because, otherwise, **making a whole scene called `ComplexNumbers` (an enormous topic) with sections named `cartesian_representation`, `polar_representation`, etc, also huge topics by themselves which would very easily span multiple minutes, is highly impractical and unwieldly.**

You mention that you need to reuse code. It's a common need, and in that case it's better to define custom classes or functions for the things we need to reuse. The same group of Mobjects over and over again? Define a new Mobject or Group for that. The same sequence of animations? You could either define an external function to be reused in your scenes, or make all your scenes inherit from a custom parent base scene class instead, with an already predefined method which does what you need.

What about sections? Well, for the `PolarRepresentation` scene, some possible sections could be much smaller fragments like:
```
show_plane_and_number
trigonometry
show_formulas
show_properties
scale_and_rotate_plane
...
```
Of course, if you do that with **every** scene, you'll end up with an enormous amount of videos in your folder, and finding a specific video will get a bit harder. But the good thing is: they're sequentially ordered, they have numbers in their filenames to help identify them because of the logical, temporal order in which they were shown on scene.
Hopefully now it makes more sense to have them numbered. That **does NOT** mean we should _not_ add the section name in the file: I think it's a good idea to have **both** of them.

With that said...

## 2. The ideas we've been discussing

There have been 4 ideas discussed here. I'll label them as **[A], [B], [C], [D]**:
| Label | When a section has a custom name... | File name format |
| --- | --- | --- |
| **[A]** | ...put only its number anyways on the file | `scene_num.ext` |
| **[B]** | ...get rid of its number and replace it with the name | `scene_name.ext` |
| **[C]** | ...use both: the number and the name, in that order | `scene_num_name.ext` |
| **[D]** | ...use both: the name and the number, in that order | `scene_name_num.ext` |

So, recapping:

- You say that **[A]** is not really clear (and I agree on that statement _alone_).
- But then you propose **[B]** instead.
- behackl said **[B]** is not a good idea, and that **preserving the number along with the name** (instead of getting rid of the number) to keep the file sorting would be nicer (which I agree with).
  - Both **[C]** and **[D]** fit at first...
  - ...but given that a key idea is "to keep the file sorting", that leaves us with **[C]**.
  - I agree wholeheartedly with him, and I think that **[C]** is the best of all 4 ideas.

Because I explained what scenes and sections are more meant to be, hopefully it makes more sense why it's good to keep the section numbering, which would answer your question:
> why it would be a problem to have unsorted files?
At the end, even when you have a file with multiple scenes you would have them unsorted.

For the same example, it would be good to have this, which fits **[C]**:
```
PolarRepresentation_0001_show_plane_and_number.mov
PolarRepresentation_0002_trigonometry.mov
PolarRepresentation_0003_show_formulas.mov
PolarRepresentation_0004_show_properties.mov
PolarRepresentation_0005_scale_and_rotate_plane.mov
```
- Because sections have a much more limited scope, and they have a time order ingrained into them, it would not be good to get rid of the numbers and let **[B]** happen instead:
```
PolarRepresentation_scale_and_rotate_plane.mov
PolarRepresentation_show_formulas.mov
PolarRepresentation_show_plane_and_number.mov
PolarRepresentation_show_properties.mov
PolarRepresentation_trigonometry.mov
```
  Yeah, I can locate them... but it can quickly get messier.
- On the other hand, I agree that having only the numbers (**[A]**) is not clear enough:
```
PolarRepresentation_0001.mov
PolarRepresentation_0002.mov
PolarRepresentation_0003.mov
PolarRepresentation_0004.mov
PolarRepresentation_0005.mov
```
  although I do give it the advantage that, because sections are so small, have a limited scope and must be ordered, chances are I would still remember which one did what ("oh, so on the 2nd section I was showing the formulas to obtain the modulus and angle of a complex number").

But why not have the best of both worlds? That's what my first proposal represents: the idea **[C]**.

> MrDiver: Because if you read them back from disk it becomes a problem of sorting the files by name
> ubaldot: This sounds more like a developer issue than a user issue.

**[B]** is an issue for **everyone and everything** that wants to work with sections in any more complex manner.

Suppose that I want to develop a specific software which takes a scene as input and displays a timeline where I can graphically have its animations displayed, along with their duration and logical order, in a way that I can edit some code in the scene and only have that specific part rerendered, being redirected in the debugger to that instant in time to see my changes immediately, as well as how it fits with the rest of the project.

Eventually I would need to support sections. For that, it would be helpful if they were already chronologically ordered.

**Removing the numbers breaks that purpose and would make it more difficult for me to find the order of the sections.** I would need to find a way to map a section to its correct position in the section order. Luckily Manim already provides me with the mapping I need: the numbers.

> True that they are sorted by name in video software but this is in-fact very desirable.

It is for scenes, not so much for sections. I already explained why. It's nice to have the numbers there.

> Hence, I wouldn't say that the current implementation is to solve a user problem.

_Straw man fallacy_. MrDiver never said that "the current implementation is to solve an user problem". Instead, he argued that _removing the numbers creates another problem_.

>Adding both names and numbers could be somewhat confusing - but still way better than having just numbers and still acceptable:
```
ComplexNumbers_cartesian_representation_001.mov
ComplexNumbers_polar_representation_002.mov
ComplexNumbers_conversions_003.mov
...
```

Another straw man. What you describe is what I labeled as **[D]**. The thing is, _no one has argued in favor of **[D]**_. The argument is whether to keep or not the sorting by number, and if you notice, **[D]** does not do that at all, because the name is before the number. It would get sorted by name anyways. **In my opinion, [D] is possibly the worst of all 4 ideas**.

Compare it to the idea **[C]**. That one _does keep the sorting by number_, which I argued is nice for sections. At the same time, it is very readable by users: they are chronologically sorted, but also the number part is only 4 digits long, so the users can immediately read the name of the section and say "oh! that was what this section did" and still understand that the section was supposed to be in that specific position and not else where.

That's why I think that idea **[C]** is the best one.

> To be the devil's advocate

the alleged "devil" here is **[C]**, not **[D]**.

> what shall be the first clip to add in the storyline? ComplexNumbers_introduction_005.mov or ComplexNumbers_cartesian_representation_001.mov?

Then again, they should have been scenes instead of sections: `Introduction` and `CartesianRepresentation`.

**Comment by MrDiver** [comment]
Let's just leave it at we will have both the number and the section name in a format that was already commented in the PR.
